arch/risc-v/src/common/riscv_misaligned.c

```diff
@@ -0,0 +1,501 @@
+ * Description:
+ *   Sign extension a given bit width data
+ *
+ ****************************************************************************/
+
+static intptr_t sext(intptr_t v, uint32_t w)
+{
+  w = 8 * sizeof(intptr_t) - w;
+  return v << w >> w;
+}
+
+/****************************************************************************
+ * Name: decode_insn_compressed
+ *
+ * Description:
+ *   Try to decode a compressed instruction
+ *
+ ****************************************************************************/
+
+static bool decode_insn_compressed(uintptr_t *regs, riscv_insn_ctx_t *ctx)
+{
+#ifdef CONFIG_ARCH_RV_ISA_C
+  uint16_t in;
+  uint32_t imm;
+  riscv_insn_c_t insn;
+
+  /* Fetch instruction */
+
+  memcpy(&insn.insn, (void *)regs[REG_EPC], 2);
+
+  in = insn.insn & INSN_C_MASK;
+
+  switch (in)
+    {
+      case INSN_C_LW:
+        imm = insn.lw.imm2 << 2 | insn.lw.imm53 << 3 | insn.lw.imm6 << 6;
+        ctx->dest = (uint8_t *)&regs[REG_X8 + insn.lw.rd];
+        ctx->src  = (uint8_t *)regs[REG_X8 + insn.lw.rs1] + imm;
+        ctx->len  = 4;
+        ctx->sext = true;
+        break;
+
+      case INSN_C_LWSP:
+        imm = insn.lwsp.imm42 << 2 | insn.lwsp.imm5 << 5 |
+              insn.lwsp.imm76 << 6;
+        ctx->dest = (uint8_t *)&regs[insn.lwsp.rd];
+        ctx->src  = (uint8_t *)regs[REG_SP] + imm;
+        ctx->len  = 4;
+        ctx->sext = true;
+        break;
+
+#  ifdef CONFIG_ARCH_RV64
+      case INSN_C_LD:
+        imm = insn.ld.imm53 << 3 | insn.ld.imm76 << 6;
+        ctx->dest = (uint8_t *)&regs[REG_X8 + insn.ld.rd];
+        ctx->src  = (uint8_t *)regs[REG_X8 + insn.ld.rs1] + imm;
+        ctx->len  = 8;
+        break;
+
+      case INSN_C_LDSP:
+        imm = insn.ldsp.imm43 << 3 | insn.ldsp.imm5 << 5 |
+              insn.ldsp.imm86 << 6;
+        ctx->dest = (uint8_t *)&regs[insn.ld.rd];
+        ctx->src  = (uint8_t *)regs[REG_SP] + imm;
+        ctx->len  = 8;
+        break;
+
+      case INSN_C_SD:
+        imm = insn.sd.imm53 << 3 | insn.sd.imm76 << 6;
+        ctx->dest = (uint8_t *)regs[REG_X8 + insn.sd.rs1] + imm;
+        ctx->src  = (uint8_t *)&regs[REG_X8 + insn.sd.rs2];
+        ctx->len  = 8;
+        break;
+
+      case INSN_C_SDSP:
+        imm = insn.sdsp.imm53 << 3 | insn.sdsp.imm86 << 6;
+        ctx->dest = (uint8_t *)regs[REG_SP] + imm;
+        ctx->src  = (uint8_t *)&regs[insn.sdsp.rs2];
+        ctx->len  = 8;
+        break;
+#  endif
+
+      case INSN_C_SW:
+        imm = insn.sw.imm2 << 2 | insn.sw.imm53 << 3 | insn.sw.imm6 << 6;
+        ctx->dest = (uint8_t *)regs[REG_X8 + insn.sw.rs1] + imm;
+        ctx->src  = (uint8_t *)&regs[REG_X8 + insn.sd.rs2];
+        ctx->len  = 4;
+        break;
+
+      case INSN_C_SWSP:
+        imm = insn.swsp.imm52 << 2 | insn.swsp.imm76 << 6;
+        ctx->dest = (uint8_t *)regs[REG_SP] + imm;
+        ctx->src  = (uint8_t *)&regs[insn.swsp.rs2];
+        ctx->len  = 4;
+        break;
+
+#  ifdef CONFIG_ARCH_FPU
+#    ifdef CONFIG_ARCH_RV32
```

Comment:
let's remove the guard?

Reply:
We can't remove this guard since `c.ld` and `c.flw` are mutually exclusive, the share the same opcode, but have different behavior on rv32/rv64.

Reply:
if so, let's remove CONFIG_ARCH_FPU guard

Reply:
OK